x/poolmanager/router_test.go

```diff
@@ -3432,3 +3439,204 @@ func (s *KeeperTestSuite) TestTrackVolume() {
+			{
+				Pools: []types.SwapAmountInRoute{
+					{
+						PoolId:        fooAbcPoolId,
+						TokenOutDenom: foo,
+					},
+				},
+				TokenInAmount: twentyFiveBaseUnitsAmount,
+			},
+		}
+		totalExpectedTakerFee = sdk.NewDecFromInt(twentyFiveBaseUnitsAmount).Mul(nonZeroTakerFee)
+		osmoTakerFeeDistr     = s.App.PoolManagerKeeper.GetParams(s.Ctx).TakerFeeParams.OsmoTakerFeeDistribution
+		nonOsmoTakerFeeDistr  = s.App.PoolManagerKeeper.GetParams(s.Ctx).TakerFeeParams.NonOsmoTakerFeeDistribution
+	)
+
+	tests := map[string]struct {
+		routes            []types.SwapAmountInSplitRoute
+		tokenInDenom      string
+		tokenOutMinAmount sdk.Int
+
+		expectedTokenOutEstimate                            sdk.Int
+		expectedTakerFees                                   expectedTakerFees
+		expectedCommunityPoolBalancesDelta                  sdk.Coins // actual community pool
+		expectedCommunityPoolFeeCollectorBalanceDelta       sdk.Coins // where non whitelisted fees are staged prior to being swapped and sent to community pool
+		expectedStakingRewardFeeCollectorMainBalanceDelta   sdk.Coins // where fees are staged prior to being distributed to stakers
+		expectedStakingRewardFeeCollectorTxfeesBalanceDelta sdk.Coins // where tx fees are initially sent, swapped, and then sent to main fee collector
+
+		expectError error
+	}{
+		"native denom taker fee": {
+			routes:            quoteNativeDenomRoute,
+			tokenInDenom:      uosmo,
+			tokenOutMinAmount: sdk.OneInt(),
+
+			expectedTokenOutEstimate: twentyFiveBaseUnitsAmount,
+			expectedTakerFees: expectedTakerFees{
+				communityPoolQuoteAssets:    sdk.NewCoins(),
+				communityPoolNonQuoteAssets: sdk.NewCoins(),
+				stakingRewardAssets:         sdk.NewCoins(sdk.NewCoin(uosmo, totalExpectedTakerFee.Mul(osmoTakerFeeDistr.StakingRewards).TruncateInt())),
+			},
+			// full native denom set in the main fee collector addr
+			expectedStakingRewardFeeCollectorMainBalanceDelta: sdk.NewCoins(sdk.NewCoin(uosmo, totalExpectedTakerFee.Mul(osmoTakerFeeDistr.StakingRewards).TruncateInt())),
+		},
+		"quote denom taker fee": {
+			routes:            quoteQuoteDenomRoute,
+			tokenInDenom:      foo,
+			tokenOutMinAmount: sdk.OneInt(),
+
+			expectedTokenOutEstimate: twentyFiveBaseUnitsAmount,
+			expectedTakerFees: expectedTakerFees{
+				communityPoolQuoteAssets:    sdk.NewCoins(sdk.NewCoin(foo, totalExpectedTakerFee.Mul(nonOsmoTakerFeeDistr.CommunityPool).TruncateInt())),
+				communityPoolNonQuoteAssets: sdk.NewCoins(),
+				stakingRewardAssets:         sdk.NewCoins(sdk.NewCoin(foo, totalExpectedTakerFee.Mul(nonOsmoTakerFeeDistr.StakingRewards).TruncateInt())),
+			},
+			// since foo is whitelisted token, it is sent directly to community pool
+			expectedCommunityPoolBalancesDelta: sdk.NewCoins(sdk.NewCoin(foo, totalExpectedTakerFee.Mul(nonOsmoTakerFeeDistr.CommunityPool).TruncateInt())),
+			// foo swapped for uosmo, uosmo sent to main fee collector, 1 uosmo diff due to slippage from swap
```

Comment:
My only note on this PR is making sure this slippage is right, but if we're sure of that, this looks good